### PR TITLE
MONGOID-5474 add readonly! and raise on save/update

### PR DESCRIPTION
**Pull request by Neilshweky**
Flip feature flag on 8.1: #5456

### CODE REVIEW

File: spec/mongoid/config_spec.rb [hunk 1]
```diff
@@ -349,6 +349,13 @@
     it_behaves_like "a config option"
   end
 
+  context 'when setting the legacy_attributes option in the config' do
+    let(:option) { :legacy_readonly }
+    let(:default) { true }
```
Comment (Neilshweky) [inline spec/mongoid/config_spec.rb:354]:
```suggestion
    let(:default) { false }
```

File: lib/mongoid/config.rb [hunk 2]
```diff
@@ -127,6 +127,16 @@ module Config
     # always return a Hash.
     option :legacy_attributes, default: false
 
+    # When this flag is true, a document is only readonly if it has been
+    # projected using #only or #without, and readonly documents will only not
```
Comment (p-mongo) [inline lib/mongoid/config.rb:131]:
```suggestion
    # projected using #only or #without, and readonly documents will not
```

File: lib/mongoid/config.rb [hunk 3]
```diff
@@ -127,6 +127,16 @@ module Config
...
+    # be able to be deleted/destroyed. When this flag is false, a document will
```
Comment (p-mongo) [inline lib/mongoid/config.rb:132]:
```suggestion
    # be deletable nor destroyable. When this flag is false, a document will
```

File: lib/mongoid/stateful.rb [hunk 4]
```diff
@@ -95,14 +95,37 @@ def pushable?
...
+    # Flags the document as readonly. Will cause a ReadonlyDocument error to be
+    # raised if the document is attempted to be saved, updated or destroyed.
+    #
+    # @example Flag the document as readonly.
+    #   document.readonly!
+    #
+    # @return [ true | false ] true if the document was successfully marked
+    #   readonly, false otherwise.
+    #
+    # @since 4.0.0
```
Comment (p-mongo) [inline lib/mongoid/stateful.rb:107]: I don't think this is right?

**Review by p-mongo — APPROVED** [review]

File: lib/mongoid/stateful.rb (hunk 4 above)
Comment (Neilshweky) [reply lib/mongoid/stateful.rb:107]: Oops lol

File: lib/mongoid/stateful.rb [hunk 5]
```diff
@@ -95,14 +95,37 @@ def pushable?
...
 
+    # Flags the document as readonly. Will cause a ReadonlyDocument error to be
+    # raised if the document is attempted to be saved, updated or destroyed.
+    #
+    # @example Flag the document as readonly.
+    #   document.readonly!
+    #
+    # @return [ true | false ] true if the document was successfully marked
+    #   readonly, false otherwise.
+    #
```
Comment (Neilshweky) [inline lib/mongoid/stateful.rb:106]:
```suggestion
```

**Pull request merged**